include/revng/Model/NamingHelper.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ADT/RecursiveCoroutine.h"
+#include "revng/Model/ArrayType.h"
+#include "revng/Model/CABIFunctionDefinition.h"
+#include "revng/Model/EnumDefinition.h"
+#include "revng/Model/Identifier.h"
+#include "revng/Model/RawFunctionDefinition.h"
+#include "revng/Model/StructDefinition.h"
+#include "revng/Model/UnionDefinition.h"
+#include "revng/Support/CommonOptions.h"
+
+namespace model {
+
+namespace detail {
+
+struct TransparentStringHash {
+  using is_transparent = void;
+
+  std::size_t operator()(const char *Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(const std::string &Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(std::string_view Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(llvm::StringRef Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(const model::Identifier &Value) const {
+    return std::hash<std::string_view>{}(Value.str());
+  }
+};
+
+} // namespace detail
+
+class VerifyHelper;
+
+class NamingHelper {
+private:
+  using GlobalNameCacheType = std::unordered_map<std::string,
+                                                 std::string,
+                                                 detail::TransparentStringHash,
+                                                 std::equal_to<>>;
+
+  GlobalNameCacheType GlobalNameCache = {};
+
+public:
+  const NamingConfiguration &Configuration;
+
+private:
+  NamingHelper(const NamingConfiguration &Configuration) :
+    Configuration(Configuration) {}
+
+  std::optional<std::string_view> isNameForbidden(std::string_view Name);
+  bool populateGlobalNameCache(VerifyHelper &VH, const Binary &Binary);
+
+public:
+  static std::optional<NamingHelper> tryMake(VerifyHelper &VH,
+                                             const Binary &Binary);
+  NamingHelper(const Binary &Binary);
```

Comment:
Push the constructors at the top.

I think the pattern for using `NamingHelper` should be:

1. Lazily call `populateGlobalNameCache` so that we can call `LLVMFunction(...)` without invoking it (if `DebugNames` is turned off).
2. When calling it lazily from the various methods, assert it's successful.
3. Drop `tryMake`. Instead, construct the regular object and, if a certain call sites expects failures, let it call `populateGlobalNameCache` proactively and check the result.

Reply:
Done!